add: blob update

dageavtobusnick opened the pull request:
<!-- Пишите **НИЖЕ** заголовков и **ВЫШЕ** комментариев, иначе ваш текст может не отобразиться. -->
<!-- В Contributing.MD вы можете найти некоторые рекомендации к оформлению пулл-реквеста. -->

## Описание
Основа -- порт блоба с https://github.com/tgstation/tgstation . Так же добавлено несколько твиков(поддержка мультиЗ, изменение принципа захвата стен и космоса) и адаптаций кода под работу с несколькими блобами, а так же с имеющимся режимом и ивентом.
Список изменений геймплея: https://docs.google.com/document/d/17k94EzPNn2UXjEDn77zk-n0zdcvHcrV6emcQkqJULQI/edit?usp=sharing
Портировано: несколько трейтов, компонентов, подсистема дыма. Все это необходимо для работы портированного кода. Отрефакторена механика рикошета. Добавлено несколько кнопочек для админов(замена штамма, включение бесконечных очков)
Весь текст при порте был переведен/адаптирован.
<!-- Опишите, что делает ваш ПР. Документировать каждую деталь не требуется, просто укажите основные изменения. -->
## Демонстрация изменений
https://youtu.be/abSmbf3JFgg

## Ссылка на предложение/Причина создания ПР
https://discord.com/channels/617003227182792704/1302986742361624627/1305226682944192552
<!-- Здесь оставьте ссылку на сообщение в #отчеты-по-предложениям, чтобы подтвердить, что ваше предложение одобрено. Либо укажите, почему этот ПР должен пройти без предложки. -->
<!-- Пример ссылки: https://discord.com/channels/617003227182792704/755125334097133628/ID-сообщения -->
## Тесты
Неколько запусков игры, в которых были проверены работы блоба/мобов/штаммов и тд. ТК пр большой, нужен будет тестмердж.
<!-- Как вы тестировали свой код. Ревьюеру будет проще, если он будет знать какие действия вы совершали, проверяя свой ПР. -->

dageavtobusnick committed:
part_1

dageavtobusnick committed:
part_2. port ended. Game running and blob works

dageavtobusnick committed:
some small improves, multiz blob support, bugfixes and translates

dageavtobusnick committed:
Merge remote-tracking branch 'upstream/master220' into blob_update_part_2

dageavtobusnick committed:
new blob storm effect + clusterbang nerf

dageavtobusnick committed:
Merge remote-tracking branch 'upstream/master220' into blob_update_part_2

dageavtobusnick committed:
fixes

dageavtobusnick committed:
Text fixes + nukecodes treshhold + blobstorm color

dageavtobusnick committed:
lint fix

dageavtobusnick committed:
text_fixes+blob mobs say

dageavtobusnick committed:
Merge remote-tracking branch 'upstream/master220' into blob_update_part_2

dageavtobusnick committed:
fix

dageavtobusnick committed:
intercepts text fixes

Den109G commented on code/__HELPERS/_logging.dm:
```diff
@@ -396,3 +396,16 @@ GLOBAL_PROTECT(log_end)
+		return "(INVALID LOCATION)"
+
+	var/turf/T = A
+	if (!istype(T))
```
```suggestion
	if(!istype(T))
```

Den109G commented on code/__HELPERS/chat.dm:
```diff
@@ -0,0 +1,15 @@
+/proc/send_to_observers(message, source)
+	var/list/all_observers = GLOB.dead_player_list + GLOB.current_observers_list
+	for(var/mob/observer as anything in all_observers)
+		if (isnull(source))
```
```suggestion
		if(isnull(source))
```

Den109G commented on code/controllers/subsystem/fluids.dm:
```diff
@@ -0,0 +1,244 @@
+ * Ensures that the subsystem's fire wait evenly splits the spread and effect waits.
+ */
+/datum/controller/subsystem/fluids/proc/initialize_waits()
+	if (spread_wait <= 0)
```
```suggestion
	if(spread_wait <= 0)
```

Den109G commented on code/controllers/subsystem/fluids.dm:
```diff
@@ -0,0 +1,244 @@
+	if (spread_wait <= 0)
+		WARNING("[src] has the invalid spread wait [spread_wait].")
+		spread_wait = 1 SECONDS
+	if (effect_wait <= 0)
```
```suggestion
	if(effect_wait <= 0)
```

Den109G commented on code/controllers/subsystem/fluids.dm:
```diff
@@ -0,0 +1,244 @@
+			to_spread.spread(seconds_per_tick)
+			to_spread.spread_bucket = null
+
+		if (MC_TICK_CHECK)
```
```suggestion
		if(MC_TICK_CHECK)
```

Den109G commented on code/controllers/subsystem/fluids.dm:
```diff
@@ -0,0 +1,244 @@
+			to_process.effect_bucket = null
+			to_process.datum_flags &= ~DF_ISPROCESSING
+
+		if (MC_TICK_CHECK)
```
```suggestion
		if(MC_TICK_CHECK)
```

Den109G commented on code/controllers/subsystem/fluids.dm:
```diff
@@ -0,0 +1,244 @@
+		var/obj/effect/particle_effect/fluid/to_process = currentrun[currentrun.len]
+		currentrun.len--
+
+		if (QDELETED(to_process) || to_process.process(seconds_per_tick) == PROCESS_KILL)
```
```suggestion
		if(QDELETED(to_process) || to_process.process(seconds_per_tick) == PROCESS_KILL)
```

Den109G commented on code/controllers/subsystem/fluids.dm:
```diff
@@ -0,0 +1,244 @@
+ * - [node][/obj/effect/particle_effect/fluid]: The node to queue to spread.
+ */
+/datum/controller/subsystem/fluids/proc/queue_spread(obj/effect/particle_effect/fluid/node)
+	if (node.spread_bucket)
```
```suggestion
	if(node.spread_bucket)
```

Den109G commented on code/controllers/subsystem/fluids.dm:
```diff
@@ -0,0 +1,244 @@
+
+	var/bucket_index = node.spread_bucket
+	spread_carousel[bucket_index] -= node
+	if (bucket_index == spread_bucket_index)
```
```suggestion
	if(bucket_index == spread_bucket_index)
```

Den109G commented on code/controllers/subsystem/fluids.dm:
```diff
@@ -0,0 +1,244 @@
+ * - [node][/obj/effect/particle_effect/fluid]: The node to start processing.
+ */
+/datum/controller/subsystem/fluids/proc/start_processing(obj/effect/particle_effect/fluid/node)
+	if (node.datum_flags & DF_ISPROCESSING || node.effect_bucket)
```
```suggestion
	if(node.datum_flags & DF_ISPROCESSING || node.effect_bucket)
```

Den109G commented on code/controllers/subsystem/fluids.dm:
```diff
@@ -0,0 +1,244 @@
+		return
+
+	effect_carousel[bucket_index] -= node
+	if (bucket_index == effect_bucket_index)
```
```suggestion
	if(bucket_index == effect_bucket_index)
```

Den109G commented on code/datums/components/blob_minion.dm:
```diff
@@ -0,0 +1,157 @@
+
+/datum/component/blob_minion/Initialize(mob/camera/blob/overmind, datum/callback/on_strain_changed)
+	. = ..()
+	if (!isminion(parent))
```
```suggestion
	if(!isminion(parent))
```

Den109G commented on code/datums/components/blob_minion.dm:
```diff
@@ -0,0 +1,157 @@
+	register_overlord(overmind)
+
+/datum/component/blob_minion/InheritComponent(datum/component/new_comp, i_am_original, mob/camera/blob/overmind, datum/callback/on_strain_changed)
+	if (!isnull(on_strain_changed))
```
```suggestion
	if(!isnull(on_strain_changed))
```

Den109G commented on code/datums/components/blob_minion.dm:
```diff
@@ -0,0 +1,157 @@
+	register_overlord(overmind)
+
+/datum/component/blob_minion/proc/register_overlord(mob/camera/blob/overmind)
+	if (isnull(overmind))
```
```suggestion
	if(isnull(overmind))
```

Den109G commented on code/datums/components/blob_minion.dm:
```diff
@@ -0,0 +1,157 @@
+	GLOB.blob_telepathy_mobs |= parent
+
+/datum/component/blob_minion/UnregisterFromParent()
+	if (!isnull(overmind))
```
```suggestion
	if(!isnull(overmind))
```

Den109G commented on code/datums/components/blob_minion.dm:
```diff
@@ -0,0 +1,157 @@
+/// Become blobpilled when we gain a mind
+/datum/component/blob_minion/proc/on_mind_init(mob/living/minion, datum/mind/new_mind)
+	SIGNAL_HANDLER
+	if (isnull(overmind) || new_mind.has_antag_datum(/datum/antagonist/blob_minion))
```
```suggestion
	if(isnull(overmind) || new_mind.has_antag_datum(/datum/antagonist/blob_minion))
```

Den109G commented on code/datums/components/blob_minion.dm:
```diff
@@ -0,0 +1,157 @@
+/// When our icon is updated, update our colour too
+/datum/component/blob_minion/proc/on_update_status_tab(mob/living/minion, list/status_items)
+	SIGNAL_HANDLER
+	if (isnull(overmind))
```
```suggestion
	if(isnull(overmind))
```

Den109G commented on code/datums/components/blob_minion.dm:
```diff
@@ -0,0 +1,157 @@
+/datum/component/blob_minion/proc/on_space_move(mob/living/minion)
+	SIGNAL_HANDLER
+	var/obj/structure/blob/blob_handhold = locate() in range(1, parent)
+	if (!isnull(blob_handhold))
```
```suggestion
	if(!isnull(blob_handhold))
```

Den109G commented on code/datums/components/connect_containers.dm:
```diff
@@ -0,0 +1,68 @@
+
+/datum/component/connect_containers/Initialize(atom/movable/tracked, list/connections)
+	. = ..()
+	if (!ismovable(tracked))
```
```suggestion
	if(!ismovable(tracked))
```

Den109G commented on code/datums/components/ghost_direct_control.dm:
```diff
@@ -0,0 +1,156 @@
+	datum/callback/after_assumed_control,
+)
+	. = ..()
+	if (!isliving(parent))
```
```suggestion
	if(!isliving(parent))
```

Den109G commented on code/datums/components/ghost_direct_control.dm:
```diff
@@ -0,0 +1,156 @@
+	var/mob/mob_parent = parent
+	LAZYADD(GLOB.mob_spawners[format_text("[initial(mob_parent.name)]")], mob_parent)
+
+	if (poll_candidates)
```
Дальше - сам. Надеюсь смысл понял.

dageavtobusnick replied:
Убрал пробелы между if-ами и скобкой

dageavtobusnick committed:
remove if spaces

pilygun commented:
К переводу вопросов нету, все понятно и четко. Апруве ведлока

dageavtobusnick committed:
Merge remote-tracking branch 'upstream/master220' into blob_update_part_2

Antoonij commented on code/__DEFINES/blob.dm:
```diff
@@ -48,9 +49,145 @@
 #define SECOND_STAGE_WARN span_userdanger("Вы чувствуете, что вот-вот лопнете.")
 
-#define isblobbernaut(M) istype((M), /mob/living/simple_animal/hostile/blob/blobbernaut)
+#define isblobbernaut(M) istype((M), /mob/living/simple_animal/hostile/blob_minion/blobbernaut)
```
перенести в исхелперы, этому тут не место

dageavtobusnick replied:
перенес

Antoonij commented on code/modules/antagonists/blob/blob_minion.dm:
```diff
@@ -0,0 +1,88 @@
+	. += span_info("Штамм вашего надразума: <b><font color=\"[blobstrain.color]\">[blobstrain.name]</b></font>!")
+	. += span_info("Штамм <b><font color=\"[blobstrain.color]\">[blobstrain.name]</b></font> [blobstrain.shortdesc ? "[blobstrain.shortdesc]" : "[blobstrain.description]"]")
+
+/datum/objective/blob_minion
+	name = "protect the blob core"
+	explanation_text = "Защищайте ядро блоба и исполняйте приказы надразумов. Любой ценой."
+	var/datum/weakref/overmind
+
+/datum/objective/blob_minion/check_completion()
+	var/mob/camera/blob/resolved_overmind = overmind.resolve()
+	if(!resolved_overmind)
+		return FALSE
+	return resolved_overmind.stat != DEAD
```
обжективы лучше перенести к обжективам блоба/другой файл

dageavtobusnick replied:
перенес

Antoonij commented on code/modules/antagonists/blob/blob_minions/blob_mob.dm:
```diff
@@ -0,0 +1,106 @@
+/mob/living/simple_animal/hostile/blob_minion/can_be_blob()
+	return FALSE
+
+/mob/living/simple_animal/hostile/blob_minion/verb/say_to_nearby_mobs()
+	set name = "Сказать окружающим"
+	set desc = "Вы скажете введенный текст окружающим вас мобам"
+
+	var/speak_text = tgui_input_text(usr, "Что вы хотите сказать?", "Сказать окружающим", null)
+
+	if(!speak_text)
+		return
+	if(stat == CONSCIOUS)
+		add_say_logs(usr, speak_text, language = "BLOB mob_say")
+		atom_say(speak_text)
```
От таких вербов надо отказываться. Сделать из этого action/innate и выдавать его в отдельном проке, что будет вызываться при инициализации/когда за моба заходит игрок

dageavtobusnick replied:
заменил на action, который выдается мобу компонентом

Antoonij commented on code/modules/antagonists/blob/blob_minions/blob_mob.dm:
```diff
@@ -0,0 +1,106 @@
+	nightvision = 8
+	lighting_alpha = LIGHTING_PLANE_ALPHA_MOSTLY_INVISIBLE
+	can_buckle_to = FALSE
+	universal_speak = TRUE //So mobs can understand them when a blob uses Blob Broadcast
```
```suggestion
	universal_speak = TRUE // So mobs can understand them when a blob uses Blob Broadcast
```

Antoonij commented on code/modules/antagonists/blob/blob_minions/blob_mob.dm:
```diff
@@ -0,0 +1,106 @@
+	icon_state = "blob_head"
+	unique_name = TRUE
+	pass_flags = PASSBLOB
+	status_flags = NONE //No throwing blobspores into deep space to despawn, or throwing blobbernaughts, which are bigger than you.
```
```suggestion
	status_flags = NONE // No throwing blobspores into deep space to despawn, or throwing blobbernaughts, which are bigger than you.
```

Antoonij commented on code/modules/antagonists/blob/blobstrains/_blobstrain.dm:
```diff
@@ -0,0 +1,172 @@
+	for(var/mob/living/blob_mob as anything in overmind.blob_mobs)
+		blob_mob.maxHealth *= max_mob_health_multiplier
+		blob_mob.health *= max_mob_health_multiplier
+		blob_mob.update_icons() //If it's getting a new strain, tell it what it does!
```
```suggestion
		blob_mob.update_icons() // If it's getting a new strain, tell it what it does!
```

dageavtobusnick committed:
Update code/modules/antagonists/blob/blob_minions/blob_mob.dm

Co-authored-by: Antoonij <42318445+Antoonij@users.noreply.github.com>

dageavtobusnick committed:
Update code/modules/antagonists/blob/blob_minions/blob_mob.dm

Co-authored-by: Antoonij <42318445+Antoonij@users.noreply.github.com>

dageavtobusnick committed:
Update code/modules/antagonists/blob/blobstrains/_blobstrain.dm

Co-authored-by: Antoonij <42318445+Antoonij@users.noreply.github.com>

dageavtobusnick committed:
fix

dageavtobusnick committed:
Merge remote-tracking branch 'upstream/master220' into blob_update_part_2

dageavtobusnick committed:
return foam subsystem(commented)

dageavtobusnick committed:
wet_status_effect

dageavtobusnick committed:
replace wet to stacking

dageavtobusnick committed:
name fix

dageavtobusnick committed:
error fix

dageavtobusnick committed:
Merge remote-tracking branch 'upstream/master220' into blob_update_part_2

dageavtobusnick committed:
redefine fix

Vladisvell commented on code/game/gamemodes/blob/blob_finish.dm:
```diff
@@ -75,9 +75,9 @@
-			text += "<br/><br/><FONT size = 2><B>Блобернаут[(blobernauts.len > 1 ? "ами были" : "ом был")]:</B></FONT>"
-			for(var/datum/mind/blob in blobernauts)
+		if(minions?.len)
+			text += "<br/><br/><FONT size = 2><B>Миньен[(minions.len > 1 ? "ами были" : "ом был")]:</B></FONT>"
```
миньен?

dageavtobusnick replied:
Да. У блоба теперь в подчинении могут быть не только разумные блобернауты, но и споры с зомби. Все они называются миньенами блоба.

Vladisvell replied:
я про орфографию (миньен-миньон)

dageavtobusnick replied:
А. Окей.

dageavtobusnick replied:
Поправлю

dageavtobusnick replied:
поправил

Vladisvell commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -28,8 +28,8 @@
-		show_message("<span class='userdanger'>Блоб атакует!</span>")
-		adjustBruteLoss(10)
+		show_message(span_userdanger("Блоб атакует!"))
+		adjustBruteLoss(10)		
```
лишние пробелы

dageavtobusnick replied:
убрал

dageavtobusnick committed:
uncomment foam + style fixes

dageavtobusnick committed:
ghost no move fix

dageavtobusnick committed:
patch 1

Исправлены баги с бессмертным зомби и зомби-невидимками.
Уменьшена скорость зомби и спор.
Уменьшен модификатор сориевого взрыва при получении урона сориевой плитки.
Уменьшен шанс расширения репликационной пены

dageavtobusnick committed:
patch 2

Слегка увеличен коэффициент урона у взрывного реагента: 0.35 to 0.4
Шанс излучения при уроне по плиткам радиоактивного блоба заменен константой в 40%
Слегка увеличен шанс расширения репликационной пены при получении урона 45 to 50%
Мелкие правки стиля кода и орфографии

dageavtobusnick committed:
patch 3

Для спор распределенных нейронов отключено 10-тисекундное голосование с оповещением. Теперь чтобы играть за спору этого реагента нужно нажать на нее с помощью лкм.
Убрано разрушение полов блобом на верхних этажах. Теперь такой пол разрушается только при разрастании сквозь уровни(рост вверх, рост вниз)
Убрано дублирование бонуса от хранилища
Споры и зомби теперь не могут таскать ничего кроме других мобов.
Убрана проверка на тач протекшн у взрывного блоба, тк у него уже есть проверка на броню от взрыва, что приводило к неадекватно низкому урону у реагента.
Репликационная пена более не расширяется от горения.

dageavtobusnick committed:
patch 4

Ядра теперь уничтожаются ядерной боеголовкой при взрыве.
Фабрика больше не может производить споры при мертвом овермайнде.

dageavtobusnick committed:
patch 5

Уменьшена длительность скользкости пола у сжатой слизи
Уменьшен шанс скользкого пола после уничтожения сжатой слизи
Добавлен слабый емп импульс при атаке по меху электромагнитной сеткой

dageavtobusnick committed:
Merge remote-tracking branch 'upstream/master220' into blob_update_part_2
